File: Golang/question/question2336.go
/**
 @author: RedCrazyGhost
 @date: 2023/11/29

2336. 无限集中的最小数字

133/135 case
**/

package question

import "sort"

type SmallestInfiniteSet struct {
	count int
	arr   []int
}

func Constructor() SmallestInfiniteSet {
	return SmallestInfiniteSet{count: 1, arr: make([]int, 0)}
}

func (this *SmallestInfiniteSet) PopSmallest() int {
	var i int
	if len(this.arr) == 0 {
		i = this.count
		this.count += 1
		return i
	}
	sort.Ints(this.arr)
	i = this.arr[0]
	this.arr = this.arr[1:]
	return i

}

func (this *SmallestInfiniteSet) AddBack(num int) {
	if num < this.count {
		this.arr = append(this.arr, num)
	}
}

File: Java/src/main/java/daily/Question2336.java
package daily;

import java.util.*;

/**
 * 2336. 无限集中的最小数字
 *
 * @author RedCrazyGhost - wenxingzhan
 * @date 2023/11/29 20:44
 **/
public class Question2336 {

    private int count;
    private TreeSet<Integer> set;

    public void SmallestInfiniteSet() {
        count=1;
        set=new TreeSet<>();
    }

    public int popSmallest() {
        if (set.isEmpty()){
            return count++;
        }
        return set.pollFirst();

    }

    public void addBack(int num) {
        if (num < count){
            set.add(num);
        }
    }
}
